File: app/src/main/java/ua/in/quireg/anothermovieapp/movieDetails/MovieActivity.java
package ua.in.quireg.anothermovieapp.movieDetails;

import android.animation.ObjectAnimator;
import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.res.TypedArray;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.WindowInsets;
import android.widget.ImageView;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.palette.graphics.Palette;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.appbar.CollapsingToolbarLayout;

import ua.in.quireg.anothermovieapp.R;
import ua.in.quireg.anothermovieapp.common.Constants;
import ua.in.quireg.anothermovieapp.common.MovieItem;
import ua.in.quireg.anothermovieapp.common.UriHelper;
import ua.in.quireg.anothermovieapp.interfaces.OnFragmentInteractionListener;
import ua.in.quireg.anothermovieapp.managers.MovieDatabaseContract;
import ua.in.quireg.anothermovieapp.managers.MovieDatabaseContract.FavouriteMovies;

public class MovieActivity extends AppCompatActivity implements OnFragmentInteractionListener {

    private static final String LOG_TAG = MovieActivity.class.getSimpleName();

    private DetailsFragment mDetailsFragment;
    private MovieItem mMovie;
    private CollapsingToolbarLayout mToolbarLayout;

    @Override
    @SuppressWarnings("ConstantConditions")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mMovie = (MovieItem) getIntent().getExtras().getSerializable(Constants.MOVIE);

        setContentView(R.layout.activity_movie_details_mine);
        final Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        AppBarLayout appBarLayout = findViewById(R.id.appbar);
        mToolbarLayout = appBarLayout.findViewById(R.id.collapsing_toolbar);

        mDetailsFragment = new DetailsFragment();
        mDetailsFragment.setArguments(getIntent().getExtras());

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.container, mDetailsFragment, DetailsFragment.TAG)
                .commit();

        if (mMovie != null) {
            if (mMovie.getTitle() != null) {
                CollapsingToolbarLayout collapsingTBL = findViewById(R.id.collapsing_toolbar);
                collapsingTBL.setTitle(mMovie.getTitle());
            }
            if (mMovie.getBackdropPath() != null) {
                loadBackdropImage(mMovie.getBackdropPath());
            }
        }
        appBarLayout.setBackgroundColor(Color.parseColor("#8C000000"));
        appBarLayout.setOnApplyWindowInsetsListener(
                new View.OnApplyWindowInsetsListener() {
                    @Override
                    public WindowInsets onApplyWindowInsets(View v, WindowInsets insets) {
                        toolbar.setPadding(
                                v.getPaddingLeft(),
                                insets.getSystemWindowInsetTop(),
                                v.getPaddingRight(),
                                v.getPaddingBottom());

                        findViewById(R.id.container).setPadding(
                                v.getPaddingLeft(),
                                v.getPaddingTop(),
                                v.getPaddingRight(),
                                insets.getSystemWindowInsetBottom());

                        TypedArray a = obtainStyledAttributes(new int[]{R.attr.actionBarSize});
                        int actionBarSize = a.getDimensionPixelSize(0, -1);
                        a.recycle();

                        mToolbarLayout.setScrimVisibleHeightTrigger(
                                insets.getSystemWindowInsetTop() + actionBarSize + 1);
                        return insets;
                    }
                });

        SyncReviewsService.startActionFetchReviews(this, String.valueOf(mMovie.getId()));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.movie_menu, menu);
        Drawable d;

        if (isFavourite(mMovie)) {
            d = getDrawable(R.drawable.ic_favorite_red_24dp);
        } else {
            d = getDrawable(R.drawable.ic_favorite_black_24dp);
        }
        menu.findItem(R.id.favorite).setIcon(d);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.favorite) {
            onFragmentMessage(
                    isFavourite(mMovie)
                            ? Constants.REMOVE_FROM_FAVOURITES
                            : Constants.ADD_TO_FAVOURITES, mMovie);
            Drawable d;
            if (isFavourite(mMovie)) {
                d = getDrawable(R.drawable.ic_favorite_red_24dp);
            } else {
                d = getDrawable(R.drawable.ic_favorite_black_24dp);
            }

            item.setIcon(d);
            return true;
        } else if (item.getItemId() == android.R.id.home) {
            onBackPressed();
        }
        return super.onOptionsItemSelected(item);
    }

    private boolean isFavourite(MovieItem item) {
        try (Cursor c = getContentResolver().query(
                MovieDatabaseContract.FavouriteMovies.buildUri(item.getId()),
                null,
                null,
                null,
                null)) {
            return c != null && c.moveToFirst();
        }
    }

    @Override
    public void onFragmentMessage(String tag, Object data) {
        MovieItem item = (MovieItem) data;
        switch (tag) {
            case Constants.ADD_TO_FAVOURITES:
                ContentValues contentValues = new ContentValues();
                contentValues.put(FavouriteMovies._ID, item.getId());
                getContentResolver().insert(FavouriteMovies.CONTENT_URI, contentValues);
                break;
            case Constants.REMOVE_FROM_FAVOURITES:
                getContentResolver().delete(
                        FavouriteMovies.buildUri(item.getId()), null, null);
                break;
        }
    }

    private void loadBackdropImage(String path) {
        Uri imageUri = UriHelper.getImageUri(path, Constants.IMAGE_SIZE_ORIGINAL);
        Log.d(LOG_TAG, "Fetching: " + imageUri.toString());
        final ImageView headerImage = findViewById(R.id.toolbar_image);

        Glide.with(this)
                .asBitmap()
                .load(imageUri)
                .addListener(new RequestListener<Bitmap>() {
                    @Override
                    public boolean onLoadFailed(@Nullable GlideException e, Object model,
                                                Target<Bitmap> target,
                                                boolean isFirstResource) {
                        if (mDetailsFragment != null) {
                            mDetailsFragment.readyShowContent();
                        }
                        return false;
                    }

                    @Override
                    public boolean onResourceReady(final Bitmap resource, Object model,
                                                   Target<Bitmap> target,
                                                   DataSource dataSource, boolean isFirstResource) {
                        if (mDetailsFragment != null) {
                            mDetailsFragment.readyShowContent();
                        }
                        headerImage.setImageBitmap(resource);
                        ObjectAnimator animator =
                                ObjectAnimator.ofFloat(headerImage, View.ALPHA,
                                        headerImage.getAlpha(), 1f);
                        animator.setDuration(600);
                        animator.start();

                        return true;
                    }
                })
                .submit();
    }

    private void playWithPalette(final Bitmap resource, final CollapsingToolbarLayout collapsingTBL,
                                 final ImageView headerImage) {
        Palette.from(Bitmap.createBitmap(resource)).generate(
                new Palette.PaletteAsyncListener() {
                    @SuppressLint("ResourceType")
                    @Override
                    public void onGenerated(Palette palette) {
//                        FloatingActionButton fab = findViewById(R.id.fab);
//                        Drawable drawable = getResources()
//                                .getDrawable(R.drawable.ic_action_star, getTheme());
//
//                        drawable.setTintMode(PorterDuff.Mode.SRC_IN);
//                        drawable.setTintList(new ColorStateList(
//                                new int[][]{
//                                        new int[]{android.R.attr.state_selected},
//                                        new int[]{-android.R.attr.state_selected}
//                                },
//                                new int[]{
//                                        Color.RED,
//                                        Color.WHITE
//                                }
//                        ));
//                        fab.setImageDrawable(drawable);
//                        fab.setBackgroundTintList(ColorStateList.valueOf(
//                                palette.getMutedColor(R.attr.colorSecondary)));
//                        fab.setRippleColor(ColorStateList.valueOf(
//                                palette.getDarkMutedColor(R.attr.colorSecondary)));

                        if (palette.getDominantSwatch() != null) {
                            collapsingTBL.setExpandedTitleColor(
                                    palette.getDominantSwatch().getTitleTextColor());
                        }

                        if (palette.getVibrantSwatch() != null) {
                            collapsingTBL.setCollapsedTitleTextColor(
                                    palette.getVibrantSwatch().getTitleTextColor());
                            collapsingTBL.setContentScrimColor(
                                    palette.getVibrantSwatch().getRgb());
                        }

                        Toolbar toolbar = findViewById(R.id.toolbar);
                        if (toolbar.getNavigationIcon() != null) {
                            toolbar.getNavigationIcon().setTint(
                                    palette.getLightVibrantColor(R.attr.editTextColor));
                        }

                        collapsingTBL.setContentScrimColor(
                                palette.getMutedColor(R.attr.colorPrimary));
                        collapsingTBL.setStatusBarScrimColor(
                                palette.getDarkMutedColor(R.attr.colorPrimaryDark));

                        headerImage.setImageBitmap(resource);
                        headerImage.setVisibility(View.VISIBLE);
                    }
                });
    }
}

File: app/src/main/java/com/anothermovieapp/movieDetails/MovieActivity.kt
package com.anothermovieapp.movieDetails

import android.animation.ObjectAnimator
import android.content.ContentValues
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.palette.graphics.Palette
import com.anothermovieapp.R
import com.anothermovieapp.common.Constants
import com.anothermovieapp.repository.Movie
import com.anothermovieapp.common.UriHelper
import com.anothermovieapp.interfaces.OnFragmentInteractionListener
import com.anothermovieapp.managers.MovieDatabaseContract.FavouriteMovies
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout

class MovieActivity : AppCompatActivity(), OnFragmentInteractionListener {
    private var mDetailsFragment: DetailsFragment? = null
    private var mMovie: Movie? = null
    private var mToolbarLayout: CollapsingToolbarLayout? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mMovie = intent.extras.getSerializable(Constants.MOVIE) as Movie
        setContentView(R.layout.activity_movie_details_mine)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
        val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
        mToolbarLayout = appBarLayout.findViewById(R.id.collapsing_toolbar)
        mDetailsFragment = DetailsFragment()
        mDetailsFragment!!.arguments = intent.extras
        supportFragmentManager.beginTransaction()
                .replace(R.id.container, mDetailsFragment!!, DetailsFragment.Companion.TAG)
                .commit()
        if (mMovie != null) {
            if (mMovie!!.title != null) {
                val collapsingTBL = findViewById<CollapsingToolbarLayout>(R.id.collapsing_toolbar)
                collapsingTBL.title = mMovie!!.title
            }
            if (mMovie!!.backdropPath != null) {
                loadBackdropImage(mMovie!!.backdropPath)
            }
        }
        appBarLayout.setBackgroundColor(Color.parseColor("#8C000000"))
        appBarLayout.setOnApplyWindowInsetsListener { v, insets ->
            toolbar.setPadding(
                    v.paddingLeft,
                    insets.systemWindowInsetTop,
                    v.paddingRight,
                    v.paddingBottom)
            findViewById<View>(R.id.container).setPadding(
                    v.paddingLeft,
                    v.paddingTop,
                    v.paddingRight,
                    insets.systemWindowInsetBottom)
            val a = obtainStyledAttributes(intArrayOf(R.attr.actionBarSize))
            val actionBarSize = a.getDimensionPixelSize(0, -1)
            a.recycle()
            mToolbarLayout.setScrimVisibleHeightTrigger(
                    insets.systemWindowInsetTop + actionBarSize + 1)
            insets
        }
        SyncReviewsService.Companion.startActionFetchReviews(this, mMovie.getId().toString())
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.movie_menu, menu)
        val d: Drawable
        d = if (isFavourite(mMovie)) {
            getDrawable(R.drawable.ic_favorite_red_24dp)
        } else {
            getDrawable(R.drawable.ic_favorite_black_24dp)
        }
        menu.findItem(R.id.favorite).icon = d
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.favorite) {
            onFragmentMessage(
                    if (isFavourite(mMovie)) Constants.REMOVE_FROM_FAVOURITES else Constants.ADD_TO_FAVOURITES, mMovie)
            val d: Drawable
            d = if (isFavourite(mMovie)) {
                getDrawable(R.drawable.ic_favorite_red_24dp)
            } else {
                getDrawable(R.drawable.ic_favorite_black_24dp)
            }
            item.icon = d
            return true
        } else if (item.itemId == android.R.id.home) {
            onBackPressed()
        }
        return super.onOptionsItemSelected(item)
    }

    private fun isFavourite(item: Movie?): Boolean {
        contentResolver.query(
                FavouriteMovies.buildUri(item.getId()),
                null,
                null,
                null,
                null).use { c -> return c != null && c.moveToFirst() }
    }

    override fun onFragmentMessage(tag: String?, data: Any?) {
        val item = data as Movie?
        when (tag) {
            Constants.ADD_TO_FAVOURITES -> {
                val contentValues = ContentValues()
                contentValues.put(FavouriteMovies._ID, item.getId())
                contentResolver.insert(FavouriteMovies.CONTENT_URI, contentValues)
            }
            Constants.REMOVE_FROM_FAVOURITES -> contentResolver.delete(
                    FavouriteMovies.buildUri(item.getId()), null, null)
        }
    }

    private fun loadBackdropImage(path: String?) {
        val imageUri = UriHelper.getImageUri(path, Constants.IMAGE_SIZE_ORIGINAL)
        Log.d(LOG_TAG, "Fetching: $imageUri")
        val headerImage = findViewById<ImageView>(R.id.toolbar_image)
        Glide.with(this)
                .asBitmap()
                .load(imageUri)
                .addListener(object : RequestListener<Bitmap?> {
                    override fun onLoadFailed(e: GlideException?, model: Any,
                                              target: Target<Bitmap?>,
                                              isFirstResource: Boolean): Boolean {
                        if (mDetailsFragment != null) {
                            mDetailsFragment!!.readyShowContent()
                        }
                        return false
                    }

                    override fun onResourceReady(resource: Bitmap?, model: Any,
                                                 target: Target<Bitmap?>,
                                                 dataSource: DataSource, isFirstResource: Boolean): Boolean {
                        if (mDetailsFragment != null) {
                            mDetailsFragment!!.readyShowContent()
                        }
                        headerImage.setImageBitmap(resource)
                        val animator = ObjectAnimator.ofFloat(headerImage, View.ALPHA,
                                headerImage.alpha, 1f)
                        animator.duration = 600
                        animator.start()
                        return true
                    }
                })
                .submit()
    }

    private fun playWithPalette(resource: Bitmap, collapsingTBL: CollapsingToolbarLayout,
                                headerImage: ImageView) {
        Palette.from(Bitmap.createBitmap(resource)).generate { palette -> //                        FloatingActionButton fab = findViewById(R.id.fab);
//                        Drawable drawable = getResources()
//                                .getDrawable(R.drawable.ic_action_star, getTheme());
//
//                        drawable.setTintMode(PorterDuff.Mode.SRC_IN);
//                        drawable.setTintList(new ColorStateList(
//                                new int[][]{
//                                        new int[]{android.R.attr.state_selected},
//                                        new int[]{-android.R.attr.state_selected}
//                                },
//                                new int[]{
//                                        Color.RED,
//                                        Color.WHITE
//                                }
//                        ));
//                        fab.setImageDrawable(drawable);
//                        fab.setBackgroundTintList(ColorStateList.valueOf(
//                                palette.getMutedColor(R.attr.colorSecondary)));
//                        fab.setRippleColor(ColorStateList.valueOf(
//                                palette.getDarkMutedColor(R.attr.colorSecondary)));
            if (palette!!.dominantSwatch != null) {
                collapsingTBL.setExpandedTitleColor(
                        palette.dominantSwatch!!.titleTextColor)
            }
            if (palette.vibrantSwatch != null) {
                collapsingTBL.setCollapsedTitleTextColor(
                        palette.vibrantSwatch!!.titleTextColor)
                collapsingTBL.setContentScrimColor(
                        palette.vibrantSwatch!!.rgb)
            }
            val toolbar = findViewById<Toolbar>(R.id.toolbar)
            if (toolbar.navigationIcon != null) {
                toolbar.navigationIcon!!.setTint(
                        palette.getLightVibrantColor(R.attr.editTextColor))
            }
            collapsingTBL.setContentScrimColor(
                    palette.getMutedColor(R.attr.colorPrimary))
            collapsingTBL.setStatusBarScrimColor(
                    palette.getDarkMutedColor(R.attr.colorPrimaryDark))
            headerImage.setImageBitmap(resource)
            headerImage.visibility = View.VISIBLE
        }
    }

    companion object {
        private val LOG_TAG = MovieActivity::class.java.simpleName
    }
}